cppcache/include/geode/Cacheable.hpp

```diff
@@ -35,10 +35,10 @@ typedef SerializablePtr CacheablePtr;
 typedef Serializable Cacheable;
 
 template <typename TVALUE>
-inline CacheablePtr createValue(const std::shared_ptr<TVALUE>& value);
+inline CacheablePtr createValue(const std::shared_ptr<TVALUE>& value) {
```

Comment:
I know you are just moving this but I really don't see the utility in this. I feel like this is a hold over from some really old magic with the old SharedPtr. All we have here now is a redefinition of dynamic cast. If we can get rid of this then let's.

Reply:
Sounds good. I'll give it a try.

Reply:
Okay, I gave it a quick try.  Region header has some interesting template type resolutions.  I can do some type trait magic, but I'm not sure it's worth it.

Here's unchanged Region signatures:

    void put(const CacheableKeyPtr& key, const CacheablePtr& value...)

    template <class KEYTYPE, class VALUETYPE>
    void put(const KEYTYPE& key, const VALUETYPE& value... )

If we call `region.put(CacheableString::create("foobar"))` this matches templated put, which uses createValue template to reinterpret CacheableStringPtr into CacheablePtr.  Ideally, it seems like instead we'd want the call to match the non-templated signature.  In order to do that I can change the signatures to something like follows:

    template <class KEYTYPE, class VALUETYPE>
    inline void put(
      const typename std::enable_if<is_integral_or_char_ptr<KEYTYPE>::value>::type& key = 0,
      const typename std::enable_if<is_integral_or_char_ptr<VALUETYPE>::value>::type& value = 0,
      const SerializablePtr& arg = nullptr) {

And then we'd have to implement `is_integral_or_char_ptr`, which should be trivial.  Is this worth it?  Did I miss something?

Reply:
We can correct this in a different ticket. How about you open one to address the ambiguity here.

Reply:
Opened ticket here: https://issues.apache.org/jira/browse/GEODE-3972